### PR TITLE
Add openapi tags to Instance,Binding types

### PR DESCRIPTION
**Pull request by juanvallejo**
This patch would allow this PR to work: https://github.com/kubernetes/kubernetes/pull/41354

**Explanation**
By tagging api types with `+k8s:openapi-gen=x-kubernetes-display-all:true`, we would be able to use the kubectl discovery client to identify the resources that should be printed as part of the "all" alias in `kubectl get all`.

**Why**

Right now, `kubectl get all` fails with the error `the server doesn't have a resource type "pods"` every time, due to `k8s.io/kubernetes/pkg/kubectl/cmd/util/shortcut_restmapper.go` defaulting to a [hardcoded list of resources](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/shortcut_restmapper.go#L106-L115) any time it attempts to lookup a resource alias (which happens any time a resource is retrieved using `kubectl get...`).

**Why not just make `shortcut_restmapper.go` use the list of discovered resources from the server every time, instead of defaulting to using the hard-coded list in kubectl?**

Right now that is what the [linked PR does](https://github.com/kubernetes/kubernetes/pull/41354), however defaulting to a list of resources returned by the discovery client would involve filtering such list in order to conform to the [definition of "all"](https://github.com/kubernetes/community/blob/master/contributors/devel/kubectl-conventions.md#rules-for-extending-special-resource-alias---all). That is not possible to do with the default information returned by the openapi spec. Hence why using this PR to add tags to all of resources we intend to return as part of `kubectl get all` is the proposed way to move forward. This would allow us to lookup these tags with kubectl and use those resouces, or keep current behavior of defaulting to the [hardcoded list of resources](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/util/shortcut_restmapper.go#L106-L115) in `shortcut_restmapper.go` for backwards compatibility.

cc @pwittrock @pmorie @fabianofranz

### CODE REVIEW

**Comment by juanvallejo** [comment]
**Why am I including a copy of `openapi-gen` in this repo, rather than updating the version of `k8s.io/kubernetes` that we are vendoring, to support tagging?**

I tried, turns out this would break a few packages due to `apimachinery` no longer being under `k8s.io/kubernetes/pkg/apimachinery`, but under `k8s.io/kubernetes/staging/src/apimachinery`. If you have a painless / simple way of making this work, I would be more than happy to update this PR to do that instead :)

**Comment by MHBauer** [comment]
Is there a change made to openapi-gen for our repo and purpose, or is this strictly an upstream copy?
If it is more than a copy, I would like that change split out as a separate commit as well.

I accept the explanation for needing a copy. 

Pending that and a working build, it's lgtm. I will re-review as necessary.

**Comment by juanvallejo** [comment]
> Is there a change made to openapi-gen for our repo and purpose, or is this strictly an upstream copy?
If it is more than a copy, I would like that change split out as a separate commit as well.

it's strictly an updated upstream copy

**Comment by MHBauer** [comment]
What do you mean by 'updated'?

Can we get the kubernetes/kubernetes sha-ish that was used as the base?

**Comment by juanvallejo** [comment]
> What do you mean by 'updated'?

The copy in this PR, uses the openapi-gen changes introduced in [317376de307407b377753c4539be607400f43115](https://github.com/kubernetes/kubernetes/commit/317376de307407b377753c4539be607400f43115) which allows for tagging. The copy found in our vendored version of kube, is from commit [6b9a9442855d16e0ff0c02ab0c82b1c1665cccaf](6b9a9442855d16e0ff0c02ab0c82b1c1665cccaf)

**Comment by MHBauer** [comment]
Yup, looks good.

There are some strange whitespace changes. But it is otherwise an exact copy.

**Comment by MHBauer** [comment]
The EDIT is what is clear to me. Thank you.

File: cmd/libs/go2idl/openapi-gen/.import-restrictions [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+        "k8s.io/kubernetes/cmd/libs/go2idl",
+        "k8s.io/apimachinery/pkg/openapi",
+        "k8s.io/gengo",
+        "k8s.io/kubernetes/third_party",
+        "k8s.io/apimachinery/third_party",
+        "k8s.io/apimachinery/pkg/util/sets"
+      ]
+    }
+  ]
+}
```
Comment (duglin) [inline cmd/libs/go2idl/openapi-gen/.import-restrictions]: can you fix the trailing weird chats here?

Comment (duglin) [inline cmd/libs/go2idl/openapi-gen/.import-restrictions]: and in the other files

Comment (duglin) [inline cmd/libs/go2idl/openapi-gen/.import-restrictions]: oh, its missing a new line at the end.  Let's fix that - even github knows that's just wrong  :-)

File: cmd/libs/go2idl/openapi-gen/generators/openapi_test.go [hunk 2]
```diff
@@ -0,0 +1,351 @@
+/*
```
Comment (MHBauer) [inline cmd/libs/go2idl/openapi-gen/generators/openapi_test.go]: can we nuke this file entirely? We don't need upstream tests.

File: cmd/libs/go2idl/openapi-gen/.import-restrictions (hunk 1 above)
Comment (juanvallejo) [inline cmd/libs/go2idl/openapi-gen/.import-restrictions]: @duglin 
> oh, its missing a new line at the end

Thanks, done!

**Comment by krancour** [comment]
How important is it to be discoverable _before_ we have a working product? I'd argue not very.

Shortly after we have this actually working, I'd anticipate that we rebase against stable upstream k8s. (@pmorie correct me if that's wrong.) When we do so, won't we get this for free?

Seems better to wait and get this for free in a few weeks than do apply this stop-gap now, have it not actually buy us anything since this project doesn't _work_ yet, and then have to remember to remove that band-aid later.

**Comment by MHBauer** [comment]
@krancour As i understand this, it is less discoverability and more having kubectl work in a pleasing fashion.

**Comment by pmorie** [comment]
@juanvallejo is working on other stuff now, so I will reopen this once our PR backlog is down.

**Comment by duglin** [comment]
@pmorie should we close this for now?

**Comment by pmorie** [comment]
Yeah - I'm going to rebase and resubmit this PR today.

**Pull request closed without merging**